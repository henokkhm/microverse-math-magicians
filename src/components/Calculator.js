import { useState } from 'react';
import styles from '../styles/Calculator/Calculator.module.css';
import Button from './Button';
import ResultBar from './ResultBar';
import calculate from '../logic/calculate';

function Calculator() {
  const [state, setState] = useState({});

  const handleClick = (value) => {
    const newState = calculate(state, value);
    setState(newState);
  };

  return (
    <section>
      <div className={styles.calculator}>
        <ResultBar />
        <div className={styles.row}>
          <Button value="AC" />
          <Button value="+/-" />
          <Button value="%" />
          <Button value="รท" accent />
        </div>
        <div className={styles.row}>
          <Button value="7" />
          <Button value="8" />
          <Button value="9" />
          <Button value="x" accent />
        </div>
        <div className={styles.row}>
          <Button value="4" />
          <Button value="5" />
          <Button value="6" />
          <Button value="-" accent />
        </div>
        <div className={styles.row}>
          <Button value="1" />
          <Button value="2" />
          <Button value="3" />
          <Button value="+" accent />
        </div>
        <div className={styles.row}>
          <Button value="0" wide />
          <Button value="." />
          <Button value="=" accent />
        </div>
      </div>
    </section>
  );
}

export default Calculator;
