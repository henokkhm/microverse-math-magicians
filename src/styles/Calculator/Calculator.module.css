.calculator {
  display: flex;
  flex-direction: column;
  width: max-content;
  gap: 1px;
  background-color: rgb(225, 224, 224);
}

.row {
  display: flex;
  gap: 1px;
  width: max-content;
}
