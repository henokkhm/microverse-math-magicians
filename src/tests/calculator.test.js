import renderer from 'react-test-renderer';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Calculator from '../components/Calculator';

describe('The calculator component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('adds one and two correctly', async () => {
    // ARRANGE
    render(<Calculator />);
    const one = await screen.findByTestId('btn-1');
    const plus = await screen.findByTestId('btn-+');
    const two = await screen.findByTestId('btn-2');
    const equals = await screen.findByTestId('btn-=');

    // ACT
    await act(async () => {
      await userEvent.click(one);
    });

    await act(async () => {
      await userEvent.click(plus);
    });

    await act(async () => {
      await userEvent.click(two);
    });

    await act(async () => {
      await userEvent.click(equals);
    });

    // ASSERT
    const calculationResult = await screen.findByTestId('calculation-result');
    expect(calculationResult).toHaveTextContent('3');
  });

  test('multiply one and two correctly', async () => {
    // ARRANGE
    render(<Calculator />);
    const one = await screen.findByTestId('btn-1');
    const multiply = await screen.findByTestId('btn-x');
    const two = await screen.findByTestId('btn-2');
    const equals = await screen.findByTestId('btn-=');

    // ACT
    await act(async () => {
      await userEvent.click(one);
    });

    await act(async () => {
      await userEvent.click(multiply);
    });

    await act(async () => {
      await userEvent.click(two);
    });

    await act(async () => {
      await userEvent.click(equals);
    });

    // ASSERT
    const calculationResult = await screen.findByTestId('calculation-result');
    expect(calculationResult).toHaveTextContent('2');
  });

  test('subtract one from two correctly', async () => {
    // ARRANGE
    render(<Calculator />);
    const one = await screen.findByTestId('btn-1');
    const subtract = await screen.findByTestId('btn--');
    const two = await screen.findByTestId('btn-2');
    const equals = await screen.findByTestId('btn-=');

    // ACT
    await act(async () => {
      await userEvent.click(two);
    });

    await act(async () => {
      await userEvent.click(subtract);
    });

    await act(async () => {
      await userEvent.click(one);
    });

    await act(async () => {
      await userEvent.click(equals);
    });

    // ASSERT
    const calculationResult = await screen.findByTestId('calculation-result');
    expect(calculationResult).toHaveTextContent('1');
  });

  test('divide nine by three correctly', async () => {
    // ARRANGE
    render(<Calculator />);
    const nine = await screen.findByTestId('btn-9');
    const divide = await screen.findByTestId('btn-รท');
    const three = await screen.findByTestId('btn-3');
    const equals = await screen.findByTestId('btn-=');

    // ACT
    await act(async () => {
      await userEvent.click(nine);
    });

    await act(async () => {
      await userEvent.click(divide);
    });

    await act(async () => {
      await userEvent.click(three);
    });

    await act(async () => {
      await userEvent.click(equals);
    });

    // ASSERT
    const calculationResult = await screen.findByTestId('calculation-result');
    expect(calculationResult).toHaveTextContent('3');
  });
});
