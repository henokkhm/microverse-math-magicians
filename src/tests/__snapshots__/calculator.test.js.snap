// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The calculator component renders correctly 1`] = `
<section
  data-testid="calculator-content"
>
  <div
    className="calculator"
  >
    <div
      className="resultBar"
      data-testid="calculation-result"
    >
      0
    </div>
    <div
      className="row"
    >
      <button
        className="button  "
        data-testid="btn-AC"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="button  "
        data-testid="btn-+/-"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="button  "
        data-testid="btn-%"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="button  accent"
        data-testid="btn-รท"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="button  "
        data-testid="btn-7"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="button  "
        data-testid="btn-8"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="button  "
        data-testid="btn-9"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="button  accent"
        data-testid="btn-x"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="button  "
        data-testid="btn-4"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="button  "
        data-testid="btn-5"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="button  "
        data-testid="btn-6"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="button  accent"
        data-testid="btn--"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="button  "
        data-testid="btn-1"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="button  "
        data-testid="btn-2"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="button  "
        data-testid="btn-3"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="button  accent"
        data-testid="btn-+"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="button wide "
        data-testid="btn-0"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="button  "
        data-testid="btn-."
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="button  accent"
        data-testid="btn-="
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</section>
`;
